packages/databricks-sdk-js/src/config/MetadataServiceCredentials.ts

```diff
@@ -0,0 +1,146 @@
+ *
+ * The MUST return 4xx response if the "X-Databricks-Metadata-Version" header
+ * is not set or set to a version that the server doesn't support.
+ *
+ * The server MUST guarantee stable sessions per URL path. That is, if the
+ * server returns a token for a Host on a given URL path, it MUST continue to return
+ * tokens for the same Host.
+ *
+ * The server MUST return a 4xx response if the Host passed in the "X-Databricks-Host"
+ * header doesn't match the token.
+ *
+ * The server is expected to return a JSON response with the following fields:
+ *
+ * - access_token: The requested access token.
+ * - token_type: The type of token, which is a "Bearer" access token.
+ * - expires_on: Unix timestamp in seconds when the access token expires.
+ */
+export class MetadataServiceCredentials implements CredentialProvider {
+    public name: AuthType = "metadata-service";
+
+    async configure(config: Config): Promise<RequestVisitor | undefined> {
+        if (!config.localMetadataServiceUrl || !config.host) {
+            return;
+        }
+
+        let parsedMetadataServiceUrl: URL;
+        try {
+            parsedMetadataServiceUrl = new URL(config.localMetadataServiceUrl);
+        } catch (error) {
+            throw new ConfigError(
+                `invalid auth server URL: ${config.localMetadataServiceUrl}`,
+                config
+            );
+        }
+
+        // only allow localhost URLs
+        if (
+            parsedMetadataServiceUrl.hostname !== "localhost" &&
+            parsedMetadataServiceUrl.hostname !== "127.0.0.1"
+        ) {
+            throw new ConfigError(
+                `invalid auth server URL: ${config.localMetadataServiceUrl}`,
+                config
+            );
+        }
+
+        const response = await this.makeRequest(
+            config,
+            parsedMetadataServiceUrl
+        );
+
+        if (!response) {
+            return;
+        }
+
+        const ts = this.getTokenSource(config, parsedMetadataServiceUrl);
+        return refreshableTokenProvider(ts);
+    }
+
+    private getTokenSource(config: Config, url: URL): Provider<Token> {
+        return async () => {
+            const serverResponse = await this.makeRequest(config, url);
+            if (!serverResponse) {
+                throw new ConfigError(
+                    `error fetching auth server URL: ${url}`,
+                    config
+                );
+            }
+
+            config.logger.info(
+                `Refreshed access token from local credentials server, which expires on ${new Date(
+                    serverResponse.expires_on * 1000
+                )}`
+            );
+
+            return new Token({
+                accessToken: serverResponse.access_token,
+                expiry: serverResponse.expires_on * 1000,
+            });
+        };
+    }
+
+    private async makeRequest(
+        config: Config,
+        url: URL
+    ): Promise<ServerResponse | undefined> {
```

Comment:
Since we are listening on localhost and not going over the network we won't suffer from intermittent network issues. I think it's fine to fail after the first try.